test/hotspot/jtreg/compiler/locks/TestNestedLocksElimination.java

```diff
@@ -0,0 +1,129 @@
+ * questions.
+ *
+ */
+
+/*
+ * @test
+ * @bug 8268347
+ * @summary Nested locks optimization may create unbalanced monitor enter/exit code
+ *
+ * @run main/othervm -XX:-BackgroundCompilation
+ *                   -XX:CompileCommand=dontinline,TestNestedLocksElimination::foo
+ *                   -XX:CompileCommand=dontinline,TestNestedLocksElimination::getNext
+ *                   -XX:CompileCommand=dontinline,TestNestedLocksElimination::getHolder
+ *                   TestNestedLocksElimination
+ */
+
+import java.util.LinkedList;
+
+public class TestNestedLocksElimination {
+
+    private LinkedList<char[]> buffers = new LinkedList<>();
+    private boolean complete = false;
+    private int bufferSize;
+
+    void foo(char[] ca) {
+        // Don't inline dummy method
+    }
+
+    char[] getNext(int length, int count) {
+        // Don't inline
+        if (this.buffers.isEmpty()) {
+            return new char[100];
+        }
+        char[] b = (char[]) this.buffers.getFirst();
+        if (count >= 100) {
+            this.complete = true;
+            this.buffers.clear(); // empty
+        }
+        return b;
+    }
+
+    synchronized boolean isComplete() {
+        return this.complete;
+    }
+
+    synchronized boolean availableSegment() {
+        return (buffers.isEmpty() == false);
+    }
+
+    TestNestedLocksElimination getHolder(TestNestedLocksElimination s1, TestNestedLocksElimination s2, int count) {
+        return (count & 7) == 0 ? s2 : s1;
+    }
+
+    int test(TestNestedLocksElimination s1, TestNestedLocksElimination s2, int maxToSend) {
+        boolean isComplete = true;
+        boolean availableSegment = false;
+        int size = 0;
+        int count = 0;
+        do {
+            TestNestedLocksElimination s = getHolder(s1, s2, count++);
+
+            synchronized(s) {
+                isComplete = s.isComplete();
+                availableSegment = s.availableSegment();
+            }
+
+            synchronized (this) {
+                size = 0;
+                while (size < maxToSend) {
+                    char[] b = null;
+                    synchronized(s) {
+                        b = s.getNext(maxToSend - size, count);
+                        isComplete = s.isComplete();
+                        availableSegment = s.availableSegment();
+                    }
+                    foo(b);
+                    size += b.length;
+                }
+            }
+      } while (availableSegment == true || isComplete == false);
+        return size;
+    }
+
+    public static void main(String[] args) {
+        int count = 0;
+        int n = 0;
+
+        TestNestedLocksElimination t = new TestNestedLocksElimination();
+        TestNestedLocksElimination s1 = new TestNestedLocksElimination();
+        TestNestedLocksElimination s2 = new TestNestedLocksEliminationSub();
+
+        char[] c = new char[100];
+        // warmup
```

Comment:
Usually when I see a `warmup` comment, there's a second loop
that does the real run. In this case, I see the `warmup` comment
and then the end of the test and I wonder if the real run is missing.

Reply:
Will remove. Left over from test development.